### PR TITLE
fix(nuxt3): use shared module for dynamic paths

### PR DESCRIPTION
**Pull request by danielroe**
### 🔗 Linked issue

resolves https://github.com/nuxt/framework/issues/3437

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

This started as adding tests for dynamic paths + relative assets. This uncovered a regression where `__webpack_public_path` wasn't being set because it wasn't top-of-entry.

This PR refactors the public path to import from two virtual templates:

1. a `client-config` file that mocks `#config` for isomorphic usage (and correspondingly we remove the `#config` import protection rule as imports from `#config` now work in vite)
2. a `dynamic-paths` file that mirrors https://github.com/nuxt/framework/blob/main/packages/nitro/src/runtime/app/paths.ts so helper functions are located in a central place rather than being embedded throughout the app

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: packages/nuxt3/src/core/templates.ts [hunk 1]
```diff
@@ -168,3 +168,36 @@ export const layoutTemplate: NuxtTemplate = {
...
+      nuxt.options.dev
+        ? `const appConfig = ${JSON.stringify(nuxt.options.app)}`
+        : 'const appConfig = config.app',
+
+      'export const NUXT_BASE_URL = () => appConfig.baseURL',
+      'export const NUXT_BUILD_ASSETS_DIR = () => appConfig.buildAssetsDir',
+
+      'export const NUXT_BUILD_ASSETS_URL = (...path) => joinURL(NUXT_PUBLIC_ASSETS_URL(), NUXT_BUILD_ASSETS_DIR(), ...path)',
+
+      'export const NUXT_PUBLIC_ASSETS_URL = (...path) => {',
```
Comment (pi0) [inline packages/nuxt3/src/core/templates.ts:197]: Can we please have consistent naming with nitro paths (https://github.com/nuxt/framework/blob/main/packages/nitro/src/runtime/app/paths.ts) for function names?

File: packages/vite/src/client.ts [hunk 2]
```diff
@@ -23,7 +23,8 @@ export async function buildClient (ctx: ViteBuildContext) {
     },
     resolve: {
       alias: {
-        '#build/plugins': resolve(ctx.nuxt.options.buildDir, 'plugins/client')
+        '#build/plugins': resolve(ctx.nuxt.options.buildDir, 'plugins/client'),
+        '#config': resolve(ctx.nuxt.options.buildDir, 'client-config.mjs')
```
Comment (pi0) [inline packages/vite/src/client.ts:27]: We can make it private to be like `#_config` and keep for internal usage while advocate composable version of config to be used by users for future compatibility of per-request variable config.

File: packages/nuxt3/src/core/templates.ts (hunk 1 above)
Comment (danielroe) [reply packages/nuxt3/src/core/templates.ts:197]: The reason is that we want to ensure these don't conflict with other locally scoped variables. I can prefix with `__` perhaps?

File: packages/vite/src/client.ts (hunk 2 above)
Comment (danielroe) [reply packages/vite/src/client.ts:27]: If we do that then we will also have to create a new alias to point `#_config` to the nitro config path (or add a wrapper template - `config.client.mjs` and `config.server.mjs`?).

Comment (pi0) [reply packages/vite/src/client.ts:27]: Should is fine for server and nitro usage. My main concern is app where composables are possible. We might find better way to also drop #config with nitropack.

File: packages/nuxt3/src/core/templates.ts (hunk 1 above)
Comment (pi0) [reply packages/nuxt3/src/core/templates.ts]: This is a new module right? How they might conflict?

Comment (danielroe) [reply packages/nuxt3/src/core/templates.ts]: because we are importing them by name into other files:

https://github.com/nuxt/framework/blob/a590c401d003bb01dee0feb84f0e989780ca4e3a/packages/vite/src/plugins/dynamic-base.ts#L63-L65

Comment (pi0) [reply packages/nuxt3/src/core/templates.ts]: Since treeshaking is not relavant, can we import * as _paths and use _paths.baseURL()?

File: packages/vite/src/client.ts (hunk 2 above)
Comment (pi0) [reply packages/vite/src/client.ts:27]: Same for client build (webpack and vite) #client usage shouldn't be possible. As we discussed, it is a temporary solution for base url handling

Comment (danielroe) [reply packages/vite/src/client.ts:27]: Agreed, it's a temporary solution. But I'm not sure what change you are asking for here.

File: packages/nuxt3/src/core/templates.ts (hunk 1 above)
Comment (danielroe) [reply packages/nuxt3/src/core/templates.ts]: Sure 👍

File: packages/vite/src/client.ts (hunk 2 above)
Comment (pi0) [reply packages/vite/src/client.ts:27]: Not providing #config alias which could be used by users (and we drop in the future) but an alias that indicates to be internal :)

Comment (danielroe) [reply packages/vite/src/client.ts:27]: Ah. My point is if we add this alias here in the client config, we will also have to add a similar `#_config` alias in the server config that we somehow resolve to point to the nitro `#config`. At the moment we are handling it by leaving it as an external and letting Nitro deal with it.

Comment (pi0) [reply packages/vite/src/client.ts:27]: Having #_config alias for server looks good for now 😊

File: packages/nuxt3/src/core/plugins/import-protection.ts [hunk 3]
```diff
@@ -18,7 +18,7 @@ export const vueAppPatterns = (nuxt: Nuxt) => [
   [/(^|node_modules\/)@vue\/composition-api/],
   ...nuxt.options.modules.filter(m => typeof m === 'string').map((m: string) =>
     [new RegExp(`^${escapeRE(m)}$`), 'Importing directly from module entrypoints is not allowed.']),
-  ...['#static-assets', '#static', '#config', '#assets', '#storage', '#server-middleware']
+  ...['#static-assets', '#static', '#assets', '#storage', '#server-middleware']
```
Comment (pi0) [inline packages/nuxt3/src/core/plugins/import-protection.ts:21]: I think we can revert it back:
```suggestion
  ...['#static-assets', '#static', '#config', '#assets', '#storage', '#server-middleware']
```

File: packages/nuxt3/src/core/templates.ts [hunk 4]
```diff
@@ -168,3 +168,40 @@ export const layoutTemplate: NuxtTemplate = {
     ].join('\n')
   }
 }
+
+export const clientConfigTemplate: NuxtTemplate = {
+  filename: 'config.client.mjs',
+  getContents: () => 'export default __NUXT__.config'
```
Comment (pi0) [inline packages/nuxt3/src/core/templates.ts:174]:
```suggestion
  getContents: () => 'export default window.__NUXT__.config'
```

File: packages/nuxt3/src/core/templates.ts [hunk 5]
```diff
@@ -168,3 +168,40 @@ export const layoutTemplate: NuxtTemplate = {
...
+}
+
+export const serverConfigTemplate: NuxtTemplate = {
+  filename: 'config.server.mjs',
+  getContents: () => [
+    'import config from \'#config\'',
```
Comment (pi0) [inline packages/nuxt3/src/core/templates.ts:180]:
```suggestion
    'import config from \'#_config\'',
```

File: packages/nuxt3/src/core/templates.ts [hunk 6]
```diff
@@ -168,3 +168,40 @@ export const layoutTemplate: NuxtTemplate = {
...
+    'export default config'
+  ].join('\n')
+}
+
+export const publicPathTemplate: NuxtTemplate = {
+  filename: 'dynamic-paths.mjs',
```
Comment (pi0) [inline packages/nuxt3/src/core/templates.ts:186]: Any reason not using `paths.mjs` (same as nitro) instead of `dynamic-paths.mjs`?

File: packages/webpack/src/presets/base.ts [hunk 7]
```diff
@@ -118,6 +118,11 @@ function baseAlias (ctx: WebpackConfigContext) {
     ...options.alias,
     ...ctx.alias
   }
+  if (ctx.isClient) {
+    ctx.alias['#_config'] = resolve(ctx.nuxt.options.buildDir, 'config.client.mjs')
```
Comment (pi0) [inline packages/webpack/src/presets/base.ts:122]: Refactor: only difference is in `client.` / `server.`. We could join if branches.

File: packages/nuxt3/src/app/entry.ts [hunk 8]
```diff
@@ -1,3 +1,4 @@
+import '#build/dynamic-paths.mjs'
```
Comment (pi0) [inline packages/nuxt3/src/app/entry.ts:1]: We are not having side-effects with `dynamic-paths`. Why is this required?

File: packages/nuxt3/src/core/plugins/import-protection.ts (hunk 3 above)
Comment (danielroe) [reply packages/nuxt3/src/core/plugins/import-protection.ts:21]: No, it will warn because of the virtual template we create internally.

File: packages/nuxt3/src/core/templates.ts (hunk 6 above)
Comment (danielroe) [reply packages/nuxt3/src/core/templates.ts:186]: No, that would be fine 👍

File: packages/nuxt3/src/app/entry.ts (hunk 8 above)
Comment (danielroe) [reply packages/nuxt3/src/app/entry.ts:1]: It has a side effect in webpack builder

File: packages/nuxt3/src/core/plugins/import-protection.ts (hunk 3 above)
Comment (pi0) [reply packages/nuxt3/src/core/plugins/import-protection.ts:21]: Can we use `#_config` within virtual template then? (i'm fine even if it requires an additional alias with nitro but not allow `#config` import in client. it is addictive and potentially breaking change in the future)

File: packages/nuxt3/src/app/entry.ts (hunk 8 above)
Comment (pi0) [reply packages/nuxt3/src/app/entry.ts:1]: Can you please elaborate more I can understand? The generated template seems to consume config only in order to export utilities. (even if necessary, at least adding an exaplanation comment)

Comment (danielroe) [reply packages/nuxt3/src/app/entry.ts:1]: It is injected by the plugin.

**Review by pi0 — APPROVED** [review]
